File: src/main/java/me/testaccount666/serversystem/commands/executables/pay/CommandPay.java
package me.testaccount666.serversystem.commands.executables.pay;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.userdata.money.EconomyProvider;
import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.command.Command;
import org.bukkit.entity.Player;

import java.math.BigDecimal;
import java.math.RoundingMode;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "pay")
public class CommandPay extends AbstractServerSystemCommand {

    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        if (!checkBasePermission(commandSender, "Pay.Use")) return;
        if (handleConsoleWithNoTarget(commandSender, getSyntaxPath(command), label, 1, arguments)) return;

        if (arguments.length < 2) {
            general("InvalidArguments", commandSender).syntaxPath(getSyntaxPath(command)).label(label).build();
            return;
        }

        var targetUserOptional = getTargetUser(commandSender, arguments);

        if (targetUserOptional.isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[0]).build();
            return;
        }

        var targetUser = targetUserOptional.get();
        var targetPlayer = targetUser.getPlayer();

        var isSelf = targetUser == commandSender;

        if (isSelf) {
            command("Pay.CannotPaySelf", commandSender).build();
            return;
        }

        try {
            var amount = BigDecimal.valueOf(Double.parseDouble(arguments[1]));
            amount = amount.setScale(2, RoundingMode.HALF_UP);

            var bankAccount = commandSender.getBankAccount();

            if (!bankAccount.hasEnoughMoney(amount)) {
                command("Pay.NotEnoughMoney", commandSender).target(targetPlayer.getName()).build();
                return;
            }

            bankAccount.withdraw(amount);
            targetUser.getBankAccount().deposit(amount);

            var formattedAmount = ServerSystem.Instance.getRegistry().getService(EconomyProvider.class).formatMoney(amount);


            command("Pay.Success", commandSender).target(targetPlayer.getName())
                    .postModifier(message -> message.replace("<AMOUNT>", formattedAmount)).build();

            command("Pay.SuccessOther", targetUser).target(targetPlayer.getName()).sender(commandSender.getName().get())
                    .postModifier(message -> message.replace("<AMOUNT>", formattedAmount)).build();
        } catch (NumberFormatException ignored) {
            command("Pay.InvalidAmount", commandSender).target(targetPlayer.getName()).build();
        }
    }

    @Override
    public String getSyntaxPath(Command command) {
        return "Pay";
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        return PermissionManager.hasCommandPermission(player, "Pay.Use", false);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/pay/CommandPay.kt
package me.testaccount666.serversystem.commands.executables.pay

import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.extensions.*
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.userdata.money.EconomyProvider
import org.bukkit.command.Command
import java.math.BigDecimal
import java.math.RoundingMode

@ServerSystemCommand("pay")
class CommandPay : AbstractServerSystemCommand() {
    override fun minRequiredArguments(command: Command) = 2
    override fun getUsagePermission(command: Command) = "Pay.Use"
    override fun getSyntaxPath(command: Command?) = "Pay"

    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        if (isConsoleWithNoTarget(commandSender, getSyntaxPath(command), label, 1, *arguments)) return

        val targetUser = getTargetUser(commandSender, arguments = arguments) ?: run {
            commandSender.generalMsg("PlayerNotFound") { target(arguments[0]) }
            return
        }

        val targetPlayer = targetUser.getPlayer()!!
        val isSelf = targetUser === commandSender

        if (isSelf) {
            commandSender.commandMsg("Pay.CannotPaySelf")
            return
        }

        val amount = arguments[1].toBigDecimalOrNull()?.setScale(2, RoundingMode.HALF_UP) ?: run {
            commandSender.commandMsg("Pay.InvalidAmount") { target(targetPlayer.name) }
            return
        }

        if (amount <= BigDecimal.ZERO) {
            commandSender.commandMsg("Pay.InvalidAmount") { target(targetPlayer.name) }
            return
        }

        val bankAccount = commandSender.bankAccount

        if (bankAccount.balance < amount) {
            commandSender.commandMsg("Pay.NotEnoughMoney") { target(targetPlayer.name) }
            return
        }

        bankAccount.transfer(amount, targetUser.bankAccount)
        val formattedAmount = getService<EconomyProvider>().formatMoney(amount)

        commandSender.commandMsg("Pay.Success") {
            target(targetPlayer.name)
            postModifier { it.replace("<AMOUNT>", formattedAmount) }
        }

        targetUser.commandMsg("Pay.SuccessOther") {
            target(targetPlayer.name)
            sender(commandSender.nameSafe)
            postModifier { it.replace("<AMOUNT>", formattedAmount) }
        }
    }
}
